osu.Game/Screens/Play/GameplayState.cs

```diff
@@ -39,6 +39,21 @@ public class GameplayState
         /// </summary>
         public readonly Score Score;
 
+        /// <summary>
+        /// Whether gameplay completed without the user failing.
+        /// </summary>
+        public bool HasPassed { get; set; }
+
+        /// <summary>
+        /// Whether the user failed during gameplay.
+        /// </summary>
+        public bool HasFailed { get; set; }
+
+        /// <summary>
+        /// Whether the user quit gameplay without having either passed or failed.
+        /// </summary>
+        public bool HasQuit { get; set; }
```

Comment:
Disagree. `SpectatingUserState` contains several other enum members which make no sense in the context of `GameplayState` (`Idle`, for one). If this were to happen then it would have to be a separate enum (`GameplayCompletionState`?)

Reply:
Perhaps I wasn't clear, but I *obviously* didn't mean for it to be used without it being renamed to not contain the "Spectating" prefix, and not have the `Idle` state (if feasible though, not sure if that might impact the current code).

Reply:
I've answered this as well in https://github.com/ppy/osu/pull/16827#discussion_r802238387. Combining it into this class seems like a fair idea, but it's a larger change that touches many files which I'm not comfortable in piling up in this PR.